```{r}
library(pacman)
p_load(data.table, dplyr, ggplot2, foreach, 
GenomicRanges, yaml, purrr, fs, stringr, tximport, edgeR)
outdatadir <- "output/code/prepdata/"
```

```{r}

dparameter <- list(
    readLength = 15,
adapter5Seq = "^GTTACATCTGGTAGTCAGTCTCCAAACCACACAAAAA",
adapter5Error = 0.1,
adapter3Seq = "AGATTGGAAGAGTGTCATGTAGGGAAAGAGTG",
adapter3Error = 0.1,
trimQuality = 20,
trimLength = 15,
mapQuality = 30,
distanceToCG = 3,
outdatadir = outdatadir,
outputRDS = paste0(outdatadir, "fileRDS/"),
core = 6,
path_to_humam_genome = "/scratch/store/annotations/HomoSapiens_hg19/genome/genome.fa",
path2CG = "/scratch/store/annotations/HomoSapiens_hg19/dCG.RDS",
minOverlap = 10,
NumberOfThreads = 20,
Overhang = 99,
path_to_humam_genome_star = paste0(outdatadir, "human_genome")
)
# removeFixed <- c(0, 10, 30, 50)
#  removeFixed <- c(0, 10, 30, 50, 70, 90) # -u funkcija
#  removeFixed <- c(101, 76, 51, 41, 31) # --length funkcija
```

```{r}
make_fastq <- function(sample, parameter = dparameter) {
    inFile <- paste0("./input/", sample$sampleID, ".fastq.gz")
    outFile <- parameter$outdatadir
    htmlFile <- paste0(parameter$outdatadir, sample$sampleID, "_fastqc.html")
    cmd <- paste0("fastqc -t ", parameter$core, " ", inFile, " -o ", outFile)
    if(!(file.exists(htmlFile))) {
    system(cmd) }
    sample$fastq <- inFile
    return(sample)
}
```


```{r}
cut_adapter_3 <- function(sample, parameter = dparameter) {
  inFile <- sample$fastq
  outFile <- paste0(parameter$outdatadir, "a3_", sample$sampleID, ".fastq.gz")
    cmd <- paste0("cutadapt ",  
            " -a ", parameter$adapter3Seq,  
            " -j ", parameter$core,
            " --minimum-length ", parameter$readLength,
            " -o ", outFile, " ", inFile)
    if(!(file.exists(outFile))) {
    system(cmd) }
    sample$adapter3 <- outFile
    return(sample)
 }
```

```{r}
star_index <- function(sample, parameter = dparameter) {
  outFile <- paste0(parameter$outdatadir, "index_", sample$sampleID)
  cmd <- paste0("STAR --runThreadN ", parameter$NumberOfThreads,
                      " --runMode ", 
                      "--genomeDir ", outFile,
                      " --genomeFastaFiles ", parameter$path_to_humam_genome,
                      " --sjdbOverhang", parameter$Overhang)
  if(!(file.exists(outFile))) {
    system(cmd) }
    sample$index <- outFile
    return(sample)
}

## SAM failas 
star_mapping <- function(sample, parameter = dparameter) {
  inFile <- sample$adapter3
  outFile <- paste0(parameter$outdatadir, "star_", sample$sampleID, "Aligned.out.sam")
   cmd <- paste0("STAR --runThreadN ", parameter$NumberOfThreads, 
                      " --genomeDir ", parameter$path_to_humam_genome_star,
                      " --readFilesIn ", inFile,
                      " --quantMode TranscriptomeSAM",
                      " --outFileNamePrefix ", outFile, 
                      " --readFilesCommand zcat")
  if(!(file.exists(paste0(outFile)))) {
    system(cmd) }
    sample$star <- outFile
    return(sample)
}
```


```{r}
## BAM failas sorted
star_mapping <- function(sample, parameter = dparameter) {
  inFile <- sample$adapter3
  outFile <- paste0(parameter$outdatadir, "star_", sample$sampleID, "Aligned.sortedByCoord.out.bam")
   cmd <- paste0("STAR --runThreadN ", parameter$NumberOfThreads, 
                      " --genomeDir ", parameter$path_to_humam_genome_star,
                      " --outSAMtype BAM SortedByCoordinate",
                      " --readFilesIn ", inFile,
                      " --outFileNamePrefix ", outFile, 
                      " --readFilesCommand zcat ")
  if(!(file.exists(paste0(outFile)))) {
    system(cmd) }
    sample$star <- outFile
    return(sample)
}
```

```{r}
## BAM failas ne sorted
star_mapping <- function(sample, parameter = dparameter) {
  inFile <- sample$adapter3
  outFile <- paste0(parameter$outdatadir, "star_", sample$sampleID, "Aligned.out.bam")
  outFile2 <- paste0(parameter$outdatadir, "star_", sample$sampleID)
   cmd <- paste0("STAR --runThreadN ", parameter$NumberOfThreads, 
                      " --genomeDir ", parameter$path_to_humam_genome_star,
                      " --outSAMtype BAM Unsorted --outFilterScoreMinOverLread 0.3 --outFilterMatchNminOverLread 0.3 --readFilesIn ", inFile,
                      " --outFileNamePrefix ", outFile2, 
                      " --readFilesCommand zcat ")
  if(!(file.exists(paste0(outFile)))) {
    system(cmd) }
    sample$star <- outFile
    return(sample)
}

```
"--outFilterScoreMinOverLread 0.3 
                       --outFilterMatchNminOverLread 0.3 
                        --quantMode TranscriptomeSAM"
```{r}
bam_sort <- function(sample, parameter = dparameter) {
    inFile <- sample$star
    outFile <- paste0(parameter$outdatadir, "star_", sample$sampleID, "Sorted.bam")
    cmd <- paste0("samtools sort -@ ", parameter$core, " ", inFile,
            " | samtools view -b -@ ", parameter$core, 
            " -q ", parameter$mapQuality, #######
            " -F 4 -o ", outFile, " - ") ######
    if(!(file.exists(outFile))) {
        system(cmd)}
    cmd2 <- paste0("samtools index ", outFile) 
        system(cmd2)
    sample$sort <- outFile
    return(sample)
}

convert_to_bed <- function(sample, parameter = dparameter) {
    inFile <- sample$sort
    outFile <- paste0(parameter$outdatadir, "star_", sample$sampleID, ".bed")
    cmd <- paste0("bedtools bamtobed -i ", inFile, 
            " > ", outFile)
    if(!(file.exists(outFile))) {
        system(cmd)}
    sample$beded <- outFile
    return(sample)
}


mergeLengthReads <- function(sample, parameter = dparameter) {
    inFile <- sample$beded
    outFile <- paste0(parameter$outputRDS, "mergeReads_", sample$sampleID, ".Rds")
    if(!(file.exists(outFile))) {
    lentele <- fread(inFile)
    i <- table(lentele$V4)
    goodID <- names(i[i == 1])
    gLentele <- lentele[V4 %in% goodID, ]
    setnames(gLentele, c("chr", "start", "end", "read", "mapq", "strand")) 
     dMain <- paste0("zcat ", inFile2, " | paste - - - - | awk '{print ($1), length($3)}'") %>%
    fread() %>%
    setkey(V1)
    dMain[, V1:=gsub("^@", "", V1)]
    full_table <- dMain [V1 %in% goodID] %>%
    setnames(c("read", "length")) %>%
    merge(gLentele, "read") %>%
    setkey(chr, start, strand, length)
        saveRDS(full_table, outFile)}
    sample$merge <- outFile
    return(sample)
}




```

```{r}
samplas <- yaml.load_file("input/samples.yaml")
```

```{r}
for (i in 1:length(samplas)) {
    make_fastq(samplas[[i]])  %>%
        cut_adapter_3() %>% 
        star_mapping()}

        make_fastq(samplas[[2]]) %>%
        cut_adapter_3() %>% 
        star_mapping()
        
         %>% bam_sort() %>% 
        convert_to_bed() %>% mergeLengthReads 
```

