```{r}
library(pacman)
p_load(data.table, dplyr, ggplot2, foreach, 
GenomicRanges, yaml, purrr, fs, stringr, tximport, edgeR)
outdatadir <- "output/code/prepdata/"
```

```{r}

dparameter <- list(
    readLength = 15,
adapter5Seq = "^GTTACATCTGGTAGTCAGTCTCCAAACCACACAAAAA",
adapter5Error = 0.1,
adapter3Seq = "AGATTGGAAGAGTGTCATGTAGGGAAAGAGTG",
adapter3Error = 0.1,
trimQuality = 20,
trimLength = 15,
mapQuality = 30,
distanceToCG = 3,
outdatadir = outdatadir,
outputRDS = paste0(outdatadir, "fileRDS/"),
core = 6,
path_to_humam_genome = "/scratch/store/annotations/HomoSapiens_hg19/genome/genome.fa",
path2CG = "/scratch/store/annotations/HomoSapiens_hg19/dCG.RDS",
minOverlap = 10,
NumberOfThreads = 20,
Overhang = 99,
path_to_humam_genome_star = paste0(outdatadir, "human_genome")
)
# removeFixed <- c(0, 10, 30, 50)
#  removeFixed <- c(0, 10, 30, 50, 70, 90) # -u funkcija
#  removeFixed <- c(101, 76, 51, 41, 31) # --length funkcija
```

```{r}
make_fastq <- function(sample, parameter = dparameter) {
    inFile <- paste0("./input/", sample$sampleID, ".fastq.gz")
    outFile <- parameter$outdatadir
    htmlFile <- paste0(parameter$outdatadir, sample$sampleID, "_fastqc.html")
    cmd <- paste0("fastqc -t ", parameter$core, " ", inFile, " -o ", outFile)
    if(!(file.exists(htmlFile))) {
    system(cmd) }
    sample$fastq <- inFile
    return(sample)
}

cut_adapter_3 <- function(sample, parameter = dparameter) {
  inFile <- sample$fastq
  outFile <- paste0(parameter$outdatadir, "a3_", sample$sampleID, ".fastq.gz")
    cmd <- paste0("cutadapt ",  
            " -a ", parameter$adapter3Seq,  
            " -j ", parameter$core,
            " --minimum-length ", parameter$readLength,
            " -o ", outFile, " ", inFile)
    if(!(file.exists(outFile))) {
    system(cmd) }
    sample$adapter3 <- outFile
    return(sample)
 }

star_index <- function(sample, parameter = dparameter) {
  outFile <- paste0(parameter$outdatadir, "index_", sample$sampleID)
  cmd <- paste0("STAR --runThreadN ", parameter$NumberOfThreads,
                      " --runMode ", 
                      "--genomeDir ", outFile,
                      " --genomeFastaFiles ", parameter$path_to_humam_genome,
                      " --sjdbOverhang", parameter$Overhang)
  if(!(file.exists(outFile))) {
    system(cmd) }
    sample$index <- outFile
    return(sample)
}


## BAM failas ne sorted
star_mapping <- function(sample, parameter = dparameter) {
  inFile <- sample$adapter3
  outFile <- paste0(parameter$outdatadir, "star_", sample$sampleID, "Aligned.out.bam")
  outFile2 <- paste0(parameter$outdatadir, "star_", sample$sampleID)
   cmd <- paste0("STAR --runThreadN ", parameter$NumberOfThreads, 
                      " --genomeDir ", parameter$path_to_humam_genome_star,
                      " --outSAMtype BAM Unsorted --outFilterScoreMinOverLread 0.3 --outFilterMatchNminOverLread 0.3 --readFilesIn ", inFile,
                      " --outFileNamePrefix ", outFile2, 
                      " --readFilesCommand zcat ")
  if(!(file.exists(paste0(outFile)))) {
    system(cmd) }
    sample$star <- outFile
    return(sample)
}


bam_sort <- function(sample, parameter = dparameter) {
    inFile <- sample$star
    outFile <- paste0(parameter$outdatadir, "star_", sample$sampleID, "Sorted.bam")
    cmd <- paste0("samtools sort -@ ", parameter$core, " ", inFile,
            " | samtools view -b -@ ", parameter$core, 
           " -o ", outFile, " - ") 
    if(!(file.exists(outFile))) {
        system(cmd)
    cmd2 <- paste0("samtools index ", outFile) 
        system(cmd2)}
    sample$sort <- outFile
    return(sample)
}

convert_to_bed <- function(sample, parameter = dparameter) {
    inFile <- sample$sort
    outFile <- paste0(parameter$outdatadir, "star_", sample$sampleID, ".bed")
    cmd <- paste0("bedtools bamtobed -i ", inFile, 
            " > ", outFile)
    if(!(file.exists(outFile))) {
        system(cmd)}
    sample$beded <- outFile
    return(sample)
}

mergeLengthReads <- function(sample, parameter = dparameter) {
    inFile <- sample$beded
    outFile1 <- paste0(parameter$outputRDS, "mapped_", sample$sampleID, ".Rds") ######
    outFile2 <- paste0(parameter$outputRDS, "unmapped_", sample$sampleID, ".Rds")######
    outFile3 <- paste0(parameter$outputRDS, "allReads_", sample$sampleID, ".Rds")####
    if(!(file.exists(outFile1))) {
    gLentele <- fread(inFile)
    gLentele <- gLentele[!duplicated(gLentele$V4),]
    goodID <- gLentele$V4
    #i <- table(lentele$V4)
    #goodID <- names(i[i == 1])
    #gLentele <- lentele[V4 %in% goodID, ]
    setnames(gLentele, c("chr", "start", "end", "read", "mapq", "strand")) 
     dMain <- paste0("zcat ", sample$adapter3, " | paste - - - - | awk '{print ($1), length($3)}'") %>%
    fread() %>%
    setkey(V1)
    dMain[, V1:=gsub("^@", "", V1)]
        mapped <- dMain %>% filter(V1 %in% goodID) %>%
    setnames(c("read", "length")) %>% 
    merge(gLentele, "read") %>%
    setkey(chr, start, strand, length) 
        unmapped <- dMain %>%
    anti_join(., mapped, by = "read") 
    saveRDS(mapped, outFile1)
    saveRDS(unmapped, outFile2)
    saveRDS(dMain, outFile3)}
    sample$mapped <- outFile1
    sample$unmapped <- outFile2
    sample$allReads <- outFile3
    return(sample)
}


makeDT <- function(sample, parameter = dparameter) {
    inFile1 <- sample$mapped 
    inFile2 <- sample$unmapped
    inFile3 <- sample$allReads
    outFile <- paste0(parameter$outputRDS, "fractionDT_", sample$sampleID, ".Rds")
     if(!(file.exists(outFile))) {
  mapped <- readRDS(inFile1) %>%
          select(read, mapq, length) %>% 
          mutate(is_uniq = if_else(
                  mapq == 255,
                  true = "yes",
                  false = "no"
    )) %>% select(read, length, is_uniq)
        
  unmapped <- readRDS(inFile2) %>%
          anti_join(., mapped, by = "read") %>% 
          select(read, length) %>% 
          mutate(is_uniq = "no")

  star_count <- bind_rows("yes" = mapped,
                     "no" = unmapped,
                     .id = "is_mapped")  %>% 
                mutate(
                status = case_when(
                    (is_mapped == "yes" & is_uniq == "yes") ~ "uniqmapped",
                    (is_mapped == "yes" & is_uniq == "no")  ~ "multimapped",
                    is_mapped == "no" ~ "unmapped")
                    )   %>% 
                mutate(
                interval_length = case_when(
                      length <= 30  ~ "-30",
                      length < 50 ~ "30-50",
                      length < 100  ~ "50-100",
                      length < 150 ~ "100-150",
                      length < 200 ~ "150-200",
                      length >= 200 ~ "+200")) %>% 
                count(status, interval_length) %>%
                setnames(c("status", "interval_length", "galutine_suma"))
  interval_count <- readRDS(inFile3) %>% 
                mutate(
                interval_length = case_when(
                      length <= 30 ~ "-30",
                      length < 50 ~ "30-50",
                      length < 100  ~ "50-100",
                      length < 150 ~ "100-150",
                      length < 200 ~ "150-200",
                      length >= 200 ~ "+200")) %>% 
                count(interval_length) %>% 
                setnames(c("interval_length", "pradine_suma"))
    full_fraction <- star_count %>% merge(interval_count, "interval_length") %>% 
  mutate(fraction = galutine_suma / pradine_suma ) %>% mutate(meginys = str_extract(
    sample$sampleID, "\\d{3}R"))
  saveRDS(full_fraction, outFile) }
  sample$dt <- outFile
  return(sample)
}

```

```{r}
samplas <- yaml.load_file("input/samples.yaml")
```

```{r}
for (i in 1:length(samplas)) {
    make_fastq(samplas[[i]])  %>%
        cut_adapter_3() %>% 
        star_mapping() %>% 
        bam_sort() %>%
        convert_to_bed() %>%
        mergeLengthReads()}

        make_fastq(samplas[[1]]) %>%
        cut_adapter_3() %>% 
        star_mapping() %>% bam_sort() %>% 
        convert_to_bed() %>% mergeLengthReads() %>% makeDT()
```








```{r}
outputRDS <- "output/code/prepdata/fileRDS/"
outFile <- paste0(outpurRDS, "FullTable.Rds")
        file_names <-
    fs::dir_ls(outputRDS, regexp = "fractionDT_")
    dt <- foreach(j = file_names, .combine = 'rbind') %do% {
        readRDS(j)
    }

full_fraction$interval_length <- factor(full_fraction$interval_length, 
    levels = c("-30", "30-50", "50-100", "100-150", "150-200", "+200"))
ggplot(full_fraction, aes(x = interval_length, y = fraction, color = status, group = status)) + 
    geom_line() + geom_point()
```